go/tasks/pluginmachinery/flytek8s/pod_helper.go

```diff
@@ -104,67 +107,230 @@ func UpdatePod(taskExecutionMetadata pluginsCore.TaskExecutionMetadata,
+		var ok bool
+		if primaryContainerName, ok = taskTemplate.GetConfig()[PrimaryContainerKey]; !ok {
+			return nil, nil, "", pluginserrors.Errorf(pluginserrors.BadTaskSpecification,
+				"invalid TaskSpecification, config missing [%s] key in [%v]", PrimaryContainerKey, taskTemplate.GetConfig())
+		}
+
+		// update annotations and labels
+		if taskTemplate.GetK8SPod().Metadata != nil {
+			objectMeta.Annotations = utils.UnionMaps(objectMeta.Annotations, taskTemplate.GetK8SPod().Metadata.Annotations)
+			objectMeta.Labels = utils.UnionMaps(objectMeta.Labels, taskTemplate.GetK8SPod().Metadata.Labels)
+		}
+	} else {
+		return nil, nil, "", pluginserrors.Errorf(pluginserrors.BadTaskSpecification,
+			"invalid TaskSpecification, unable to determine Pod configuration")
 	}
+
+	return podSpec, &objectMeta, primaryContainerName, nil
+}
+
+// ApplyFlytePodConfiguration updates the PodSpec and ObjectMeta with various Flyte configuration. This includes
+// applying default k8s configuration, resource requests, injecting copilot containers, and merging with the
+// configuration PodTemplate (if exists).
+func ApplyFlytePodConfiguration(ctx context.Context, tCtx pluginsCore.TaskExecutionContext, podSpec *v1.PodSpec, objectMeta *metav1.ObjectMeta, primaryContainerName string) (*v1.PodSpec, *metav1.ObjectMeta, error) {
+	taskTemplate, err := tCtx.TaskReader().Read(ctx)
+	if err != nil {
+		logger.Warnf(ctx, "failed to read task information when trying to construct Pod, err: %s", err.Error())
+		return nil, nil, err
+	}
+
+	// add flyte resource customizations to containers
 	templateParameters := template.Parameters{
-		Task:             tCtx.TaskReader(),
 		Inputs:           tCtx.InputReader(),
 		OutputPath:       tCtx.OutputWriter(),
+		Task:             tCtx.TaskReader(),
 		TaskExecMetadata: tCtx.TaskExecutionMetadata(),
 	}
-	c, err := ToK8sContainer(ctx, task.GetContainer(), task.Interface, templateParameters)
+
+	resourceRequests := make([]v1.ResourceRequirements, 0, len(podSpec.Containers))
+	var primaryContainer *v1.Container
+	for index, container := range podSpec.Containers {
+		var resourceMode = ResourceCustomizationModeEnsureExistingResourcesInRange
+		if container.Name == primaryContainerName {
+			resourceMode = ResourceCustomizationModeMergeExistingResources
+		}
+
+		if err := AddFlyteCustomizationsToContainer(ctx, templateParameters, resourceMode, &podSpec.Containers[index]); err != nil {
+			return nil, nil, err
+		}
+
+		resourceRequests = append(resourceRequests, podSpec.Containers[index].Resources)
+		if container.Name == primaryContainerName {
+			primaryContainer = &podSpec.Containers[index]
+		}
+	}
+
+	if primaryContainer == nil {
+		return nil, nil, pluginserrors.Errorf(pluginserrors.BadTaskSpecification, "invalid TaskSpecification, primary container [%s] not defined", primaryContainerName)
+	}
+
+	// add copilot configuration to primaryContainer and PodSpec (if necessary)
+	if taskTemplate.GetContainer() != nil {
```

Comment:
I understand most of this code is just refactored to a new location but why is this only happening in case of a container not a Pod task?

Reply:
Yeah, so the recent [flytekit PR](https://github.com/flyteorg/flytekit/pull/1425) does not yet address adding the `pod_template` support to `ContainerTasks`. The difficulty here is that the `DataLoadingConfig` struct is [only available on the Container message](https://github.com/flyteorg/flyteidl/blob/c60305cb0ced352cf3918f5e0e3de116271e0479/protos/flyteidl/core/tasks.proto#L197), so not on the K8sPod. This means that currently we can only support injecting copilot uploader / downloader for `Container` instances of the `TaskTemplate`.

The plan was to address this when we implement `pod_template` for the `ContainerTask`, there have been multiple community members willing to work on it immediately. This will require minor refactoring here - ie. add `DataLoadingConfig` to either `TaskTemplate` (and deprecate the field in `Container`) or `K8sPod` proto message, and update this code to use the correct value.

Alternatively, happy to keep this PR open until we add support for `pod_template` in the `ContainerTask` which will require the aforementioned updates.